thrall/app/lib/ThrallMessageConsumer.scala

```diff
@@ -83,9 +94,62 @@ class ThrallMessageConsumer(config: ThrallConfig, es: ElasticSearch, thrallMetri
   def deleteAllUsages(usage: JsValue) =
     Future.sequence( withImageId(usage)(id => es.deleteAllImageUsages(id)) )
 
-  def updateRcsRights(rights: JsValue) =
-    Future.sequence( withImageId(rights)(id => es.updateImageSyndicationRights(id, rights \ "data")) )
+  def refreshInferredSyndicationRights(imageJson: JsValue) = {
+    Future.sequence(
+      withImageId(imageJson) { id =>
+        val image = imageJson.as[Image]
+        GridLogger.info(s"upserting inferred rights", id)
+        es.updateImageSyndicationRights(id, image.syndicationRights)
+      }
+    )
+  }
+
+  def upsertSyndicationRights(rights: JsValue) = {
+    (rights \ "data").validate[SyndicationRights].fold(
+      err => {
+        val msg = s"Unable to parse message as SyndicationRights ${JsError.toJson(err).toString}"
+        GridLogger.error(msg)
+        Future.failed(SNSBodyParseError(msg))
+      },
+      rightsFromRcs => withImageId(rights) { id =>
+        es.getImage(id) map {
+          case Some(image) => {
+            GridLogger.info(s"Upserting syndication rights from RCS feed", image.id)
+            es.updateImageSyndicationRights(id, Some(rightsFromRcs))
+
+            image.userMetadata.map{edits => {
+              edits.photoshoot match {
+                case Some(photoshoot) if !rightsFromRcs.isInferred => {
+                  // Image is in a photoshoot and has real rights
+                  // Update the other images in the photoshoot
+                  es.getImagesWithInferredSyndicationRights(photoshoot, image)
+                    .map(initiateInferredRightsRefresh(_, rightsFromRcs))
+                  image
+                }
+                case _ => image
+              }
+            }}
+          }
+          case _ => {
+            GridLogger.info(s"image $id not found")
+            None
+          }
+        }
+      }
+    )
+  }
+
+  private def initiateInferredRightsRefresh(images: List[Image], rightsFromRcs: SyndicationRights) = {
```

Comment:
This introduces a lot of logic in `ThrallMessageConsumer`, it would be nice pull this out and have smaller steps. It would also make it easier to read if instead of the nested `upsertSyndicationRights` logic we'd have a for comprehension. But it's up to you how much time you want to spend on this :)